### ISSUE TITLE
clarify separation of high-level and low-level API

### ISSUE THREAD

**Issue by grahamgower** [popsim-consortium/demes-python#73]
The `DemeGraph` class provides a high-level API for constructing a demographic model. The methods on this class ensure consistency within the demographic model. They also allow omission of detail, when there is a single unambiguous interpretation (or a very sensible default). We use these methods to construct a demographic model when loading a yaml file.

In contrast, we also have the low-level "building block" classes in `demes/demes.py`, such as `Deme`, `Epoch`, `Migration`, etc. that are used by the `DemeGraph` methods. We gaurantee correctness only when they are constructed via the `DemeGraph`. It's possible to instantiate these outside of the context of a `DemeGraph`, and so these classes do not themselves provide any gaurantees about correctness.  But it is sometimes convenient to use these classes directly, in part because they don't complain when you omit information required by a strictly-correct `DemeGraph`. E.g. when converting from msprime demographies, we iterate through the `msprime.DemographyDebugger.epochs` and we can instantiate some of these objects without complete information, and add more information as it becomes available. Developers might find other uses. General users (i.e. model implementers) should not be using this low-level API.

Probably this low/high separation is mostly a matter of documention. But there might currently be some edge cases where we blur the semantics (maybe in the `Deme` class, which has an `add_epoch()` method?). We should identify these and fix the API(s).

**Comment by jeromekelleher** [comment]
I think it's mainly a documentation thing. It's probably a case where "documentation-driven-development" would help - as you try to explain the distinctions, ways to clarify and make things simpler could occur.

**Comment by grahamgower** [comment]
No longer relevant after #208.
